src/Nip17/DirectMessage.php

```diff
@@ -0,0 +1,275 @@
+
+        // Discover receiver's preferred relays
+        $receiverRelays = $this->getPreferredRelaysForPubkey($receiverPubkey);
+
+        // Get sender's preferred relays
+        $senderRelays = $this->getPreferredRelaysForPubkey($senderPubkey);
+
+        // Create the base event (kind 14)
+        $event = $this->createDirectMessageEvent($message, $receiverPubkey, $replyToId, $additionalTags);
+        $event->setSenderPubkey($senderPubkey);
+
+        // // Step 1: Create a seal (kind 13) for the event
+        $sealEvent = $this->giftWrapService->createSeal($event, $senderPrivkey, $receiverPubkey);
+
+        // // Step 2: Create a gift wrap (kind 1059) to further protect the seal
+        $receiverGiftWrap = $this->giftWrapService->createGiftWrap($sealEvent, $receiverPubkey);
+
+        // Also create a copy for the sender to keep track of sent messages
+        $senderGiftWrap = $this->giftWrapService->createGiftWrap($sealEvent, $senderPubkey);
+
+        return [
+            'seal' => $sealEvent,
+            'receiver' => $receiverGiftWrap,
+            'sender' => $senderGiftWrap,
+            'receiver_relays' => $receiverRelays,
+            'sender_relays' => $senderRelays,
+        ];
+    }
+
+    /**
+     * Decrypt a direct message from a gift wrap and seal.
+     *
+     * @param GiftWrapInterface|EventInterface|\stdClass $giftWrap The gift wrapped event to decrypt
+     * @param string $receiverPrivkey The private key of the recipient (you)
+     * @param bool $verifyRecipient Whether to verify that the gift wrap is addressed to the receiver
+     *
+     * @return array|null The decrypted event as an array, or null if decryption failed or recipient verification failed
+     */
+    public static function decryptDirectMessage(
+        $giftWrap,
+        string $receiverPrivkey,
+        bool $verifyRecipient = true,
+    ): ?array {
+        // Extract the recipient pubkey from p tag
+        $receiverPubkey = (new Key())->getPublicKey($receiverPrivkey);
+        $isAddressedToReceiver = false;
+
+        // Check that the gift wrap is addressed to the receiver via p tag
+        // Handle both Event objects and stdClass objects from relays
+        $tags = method_exists($giftWrap, 'getTags') ? $giftWrap->getTags() : ($giftWrap->tags ?? []);
+
+        foreach ($tags as $tag) {
+            if ($tag[0] === 'p' && $tag[1] === $receiverPubkey) {
+                $isAddressedToReceiver = true;
+                break;
+            }
+        }
+
+        // If verifying recipient and the gift wrap is not for this receiver, abort
+        if ($verifyRecipient && !$isAddressedToReceiver) {
+            throw new \Exception("Gift wrap is not addressed to the receiver");
+        }
+
+        try {
+            // Extract the encrypted content from the gift wrap
+            $encryptedContent = method_exists($giftWrap, 'getContent') ?
+                $giftWrap->getContent() :
+                ($giftWrap->content ?? '');
```

Comment:
Same as my previous comment for line 105.

Reply:
Same answer ;)